lib/Sema/CSDiagnostics.cpp

```diff
@@ -8187,6 +8187,50 @@ bool UnableToInferClosureReturnType::diagnoseAsError() {
   return true;
 }
 
+bool UnableToInferGenericPackElementType::diagnoseAsError() {
+
+  const auto* locator = getLocator();
+  auto path = locator->getPath();
+
+  if (path.size() > 1) {
+  
+    const auto applyArgToParamElt = (path.end() - 2)->getAs<LocatorPathElt::ApplyArgToParam>();
+    const auto packElementElt = (path.end() - 1)->getAs<LocatorPathElt::PackElement>();
+
+    if (!applyArgToParamElt)
+      return false;
+
+    auto anchor = getAnchor();
+
+    // `nil` appears as an element of generic pack params, let's record a
+    // specify contextual type for nil fix.
+    if (isExpr<NilLiteralExpr>(anchor)) {
+      emitDiagnostic(diag::unresolved_nil_literal);
+    }
+    else {
+      // unable to infer the type of an element of generic pack params
+      emitDiagnostic(diag::could_not_infer_pack_element, packElementElt->getIndex());
+    }
+
+    // emit callee side diagnostics
+    auto applyExpr = castToExpr<ApplyExpr>(locator->getAnchor());
+    if (auto* Fn = applyExpr->getFn()) {
+      if (const auto DeclRef = Fn->getReferencedDecl()) {
```

Comment:
This would only work for non-overloaded cases like you have with`bar` in all other cases you'd have to retrieve the overload selected for a given solution using `FailureDiagnostic::getOverloadChoice(ConstraintLocator *)`, it would be good to add a few test-cases with overloads too!